README
                     __  __ ___    _____  ____  ___    __    __ 
                    /#/ /#// _ |  /####/ / / / / _ |  / /   / / 
                   /#/ /#// __ | /##/   /_  _// __ | / /__ / /__
                  /#####//_/ |_|/####/   /_/ /_/ |_|/____//____/
                                       M a e m o  E d i t i o n


This is UAE4ALL emulator port for Maemo 5 Platform (ie. Nokia N900).
Maemo release is based on the source code from Chui's UAE4ALL RC2
and critical's & notaz's UAE4ALL GP2X v0.8.0.

Homepage of UAE4All Maemo Edition:
- http://github.com/smokku/uae4all


Chui's Dreamcast UAE4ALL:
- http://chui.dcemu.co.uk/uae4all.html

notaz's UAE4ALL GP2X:
- http://notaz.gp2x.de/other.php#GP2X


---------------------------------------------------------------------------
UAE4ALL RC 1 (11/03/2006)                                Chui & Fox68k 2006

UAE4ALL is an Amiga 500 emulator for the Dreamcast videogame system.

This is a branch emulation development from Experimental UAE 0.8.22.
E-UAE is open-source software and is made available under the terms of the
GNU GPL. E-UAE is based on the work of dozens of contributors including
Bernd Schmidt (the original author and maintainer of UAE), Bernie Meyer 
(the author of the x86 JIT compiler), Toni Wilen (the current maintainer of 
WinUAE), and many more.
---------------------------------------------------------------------------
UAE4ALL RC 2 (02/04/2010)                                              Chui

Some new features for Dingoo
like savestates and some better speed for sound.
---------------------------------------------------------------------------


UAE4All Emulates Commodore Amiga 500 with additional 512kB RAM expansion.

You need to put the "kick.rom" image in the MyDocs directory (MMC root).
Make sure that you have the correct kickstart image (524288 bytes).
The original kickstart 1.3 works best.

GZip compressed ADF (.ADZ) images are supported.

For more information see README.Dingoo and README.GP2X.


This is GPL'ed software... respect the license.


Menu keymapping:
<- : A, <-
-> : D, ->
^  : W, ^
v  : S, v
Cancel : Q, LALT
Load   : C, LSHIFT
X      : E, LCTRL, ENTER
Y      : X, SPACE
R      : 1, TAB
L      : 2, BACKSPACE


Keyboard mapping:
1-0    : Fn + Q-P
F1-F10 : Fn + Ctrl + Q-P
Escape : Fn + Ctrl + BACKSPACE
Keypad : Fn + Shift + Q-P ASFV.ENTER


Toggles:
StatusBar                  : Fn + S
Mouse/Joy emulation toggle : Fn + SPACE


Mouse1 / Joystick2 keymapping:
Toggle     : Fn + SPACE   (off/joy/mouse)
Directions : Arrow Keys
Button 1   : SHIFT
Button 2   : SPACE


To calibrate real joystick do a full circle around.



debian/changelog
uae4all (1rc2-15) fremantle; urgency=low

  * Added hildon-games-wrapper support.
  * Use zlib to compress savestates.

 -- Tomasz Sterna <tomek@xiaoka.com>  Mon, 10 May 2010 00:36:43 +0200

uae4all (1rc2-14) fremantle; urgency=low

  * Implemented drive number selection.
  * Use UAE m68k emulation.
  * Removed misported sprite collision handling.

 -- Tomasz Sterna <tomek@xiaoka.com>  Mon,  3 May 2010 16:52:16 +0200

uae4all (1rc2-13) fremantle; urgency=low

  * Brought back Cyclone m68k emulator.
  * Change keyboard layout to standard QWERTY when running.

 -- Tomasz Sterna <tomek@xiaoka.com>  Wed, 28 Apr 2010 17:26:30 +0200

uae4all (1rc2-12) fremantle; urgency=low

  * Fixed horizontal downscaling.

 -- Tomasz Sterna <tomek@xiaoka.com>  Tue, 27 Apr 2010 13:50:28 +0200

uae4all (1rc2-11) fremantle; urgency=low

  * Fixed HAM and EHB modes.

 -- Tomasz Sterna <tomek@xiaoka.com>  Tue, 27 Apr 2010 11:14:11 +0200

uae4all (1rc2-10) fremantle; urgency=low

  * Fixed real joystick support.
  * Added support for SIXAXIS(TM) controller.

 -- Tomasz Sterna <tomek@xiaoka.com>  Thu, 22 Apr 2010 03:13:51 +0200

uae4all (1rc2-9) fremantle; urgency=low

  * Reverted to FAME core for additional compatibility.
  * Added shortcut to disable/enable status bar.

 -- Tomasz Sterna <tomek@xiaoka.com>  Thu, 22 Apr 2010 00:45:48 +0200

uae4all (1rc2-8) fremantle; urgency=low

  * Fixed keyboard mapping with joy/mouse emulation mode.

 -- Tomasz Sterna <tomek@xiaoka.com>  Wed, 21 Apr 2010 14:08:31 +0200

uae4all (1rc2-7) fremantle; urgency=low

  * Implemented Joystick/Mouse emulation toggle
  * Changed mapping of fire/mouse buttons

 -- Tomasz Sterna <tomek@xiaoka.com>  Wed, 21 Apr 2010 13:43:32 +0200

uae4all (1rc2-6) fremantle; urgency=low

  * Added check for required KICK.ROM size

 -- Tomasz Sterna <tomek@xiaoka.com>  Tue, 20 Apr 2010 22:26:00 +0200

uae4all (1rc2-5) fremantle; urgency=low

  * Fixed sound option selection.
  * Added 'silent' sound emulation.
  * KICK.ROM missing error message more informative.
  * Proper version information in menu.

 -- Tomasz Sterna <tomek@xiaoka.com>  Tue, 20 Apr 2010 14:12:35 +0200

uae4all (1rc2-4) fremantle; urgency=low

  * Do not show hidden files and directories in file selector.
  * Fixed ESC key mapping.
  * Fixed joystick emulation.

 -- Tomasz Sterna <tomek@xiaoka.com>  Tue, 20 Apr 2010 10:46:18 +0200

uae4all (1rc2-3) fremantle; urgency=low

  * Fixed ESC key mapping.
  * Implemented stereo sound.
  * Merged uae4all-wiz 0.8.0 changes.
  * Using Cyclone for m68k processor emulation.
  * 4 FDDs support.
  * Added PAL/NTSC selector to menu.
  * Additional ARM optimisations.
  * Added SysInfo and JoyTest disk images to distribution.

 -- Tomasz Sterna <tomek@xiaoka.com>  Mon, 19 Apr 2010 15:53:07 +0200

uae4all (1rc2-2) fremantle; urgency=low

  * Updated graphics with RC2 resources.
  * Fixed joystick emulation on Maemo.
  * Extended keyboard emulation on Maemo.

 -- Tomasz Sterna <tomek@xiaoka.com>  Sun, 18 Apr 2010 23:12:19 +0200

uae4all (1rc2-1) fremantle; urgency=low

  * Merged uae4all RC2 (20100402) source from Chui:
    * savestates support
    * better sound emulation performance
    * keyboard joystick emulation
    * 7zip compressed disks support
  * Disabled SDL_DOUBLEBUFFER (it's not needed with scaler enabled)

 -- Tomasz Sterna <tomek@xiaoka.com>  Fri, 16 Apr 2010 12:18:29 +0200

uae4all (1rc1-2) fremantle; urgency=low

  * Fixed mouse position mapping.
  * Added joystick1 keyboard emulation.

 -- Tomasz Sterna <tomek@xiaoka.com>  Fri, 16 Apr 2010 01:39:14 +0200

uae4all (1rc1-1) fremantle; urgency=low

  * Initial packaging.
  * N900 Makefile.
  * Using /home/user/MyDocs as KICK.ROM location.
  * Created Boing Ball icon for UAE4All
  * Modified SDL renderer for Maemo requirements.
  * Rendering in 800x480 with 2x software scaler
    ported from DrNokSnes.
  * Modified splash screen to be generic.
  * Fixed keymapping on Maemo (KP_ENTER).

 -- Tomasz Sterna <tomek@xiaoka.com>  Wed, 14 Apr 2010 21:36:46 +0200


src/gui.cpp
 /*
  * UAE - The Un*x Amiga Emulator
  *
  * Interface to the Tcl/Tk GUI
  *
  * Copyright 1996 Bernd Schmidt
  */

#include "sysconfig.h"
#include "sysdeps.h"
#include "config.h"
#include "uae.h"
#include "options.h"
#include "gui.h"
#include "menu.h"
#include "vkbd.h"
#include "debug_uae4all.h"
#include "custom.h"
#include "memory.h"
#include "xwin.h"
#include "drawing.h"
#include "sound.h"
#include "audio.h"
#include "keybuf.h"
#include "disk.h"
#include "savestate.h"

#include <SDL.h>

#ifdef PROFILER_UAE4ALL
unsigned long long uae4all_prof_initial[UAE4ALL_PROFILER_MAX];
unsigned long long uae4all_prof_sum[UAE4ALL_PROFILER_MAX];
unsigned long long uae4all_prof_executed[UAE4ALL_PROFILER_MAX];
#endif

#ifdef DREAMCAST
#include <SDL_dreamcast.h>
#define VIDEO_FLAGS_INIT SDL_HWSURFACE|SDL_FULLSCREEN
#else
#ifdef DINGOO
#define VIDEO_FLAGS_INIT SDL_SWSURFACE
#else
#ifdef MAEMO_CHANGES
#define VIDEO_FLAGS_INIT SDL_HWSURFACE|SDL_FULLSCREEN
#else
#define VIDEO_FLAGS_INIT SDL_HWSURFACE
#endif
#endif
#endif

#ifdef DOUBLEBUFFER
#define VIDEO_FLAGS VIDEO_FLAGS_INIT | SDL_DOUBLEBUF
#else
#define VIDEO_FLAGS VIDEO_FLAGS_INIT
#endif

static char _show_message_str[40]={
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
};

int show_message=0;
char *show_message_str=(char *)&_show_message_str[0];

extern SDL_Surface *prSDLScreen;
extern struct uae_prefs changed_prefs;
extern struct uae_prefs currprefs;
extern SDL_Joystick *uae4all_joy0, *uae4all_joy1;

extern int keycode2amiga(SDL_keysym *prKeySym);
extern int uae4all_keystate[];

int emulated_mouse_speed=2;
int emulating=0;
char uae4all_image_file0[128]  = { 0, };
char uae4all_image_file1[128] = { 0, };
char uae4all_image_file2[128]  = { 0, };
char uae4all_image_file3[128] = { 0, };

int drawfinished=0;

extern int mainMenu_throttle, mainMenu_frameskip, mainMenu_sound, mainMenu_case, mainMenu_autosave, mainMenu_ntsc, mainMenu_drives;

int emulated_joystick=0;
int emulated_left=0;
int emulated_right=0;
int emulated_top=0;
int emulated_bot=0;
int emulated_button1=0;
int emulated_button2=0;
int emulated_mouse=0;
int emulated_mouse_button1=0;
int emulated_mouse_button2=0;

static void getChanges(void)
{
    m68k_speed=mainMenu_throttle;
#ifndef NO_SOUND
    if (mainMenu_sound)
    {
	if (mainMenu_sound == 1)
	{
		changed_produce_sound=2;
	}
	else
	{
		// fake it
		changed_produce_sound=1;
	}
	sound_default_evtime();
    }
    else
#endif
	changed_produce_sound=0;
    changed_gfx_framerate=mainMenu_frameskip;
}

int gui_init (void)
{
	
//Se ejecuta justo despues del MAIN
    if (prSDLScreen==NULL)
	prSDLScreen=SDL_SetVideoMode(PREFS_GFX_WIDTH,PREFS_GFX_HEIGHT,16,VIDEO_FLAGS);
    SDL_ShowCursor(SDL_DISABLE);
    SDL_JoystickEventState(SDL_ENABLE);
    SDL_JoystickOpen(0);
    if (prSDLScreen!=NULL)
    {
	emulating=0;
	vkbd_init();
	init_text(1);
	if (!uae4all_image_file0[0])
		run_mainMenu();
	quit_text();
	uae4all_pause_music();
	emulating=1;
	getChanges();
	check_all_prefs();
	reset_frameskip();
#ifdef DEBUG_FRAMERATE
	uae4all_update_time();
#endif
#ifdef PROFILER_UAE4ALL
	uae4all_prof_init();
	uae4all_prof_add("M68K");			// 0
	uae4all_prof_add("EVENTS");			// 1
	uae4all_prof_add("HSync");			// 2
	uae4all_prof_add("Copper");			// 3
	uae4all_prof_add("Audio");			// 4
	uae4all_prof_add("CIA");			// 5
	uae4all_prof_add("Blitter");			// 6
	uae4all_prof_add("Vsync");			// 7
	uae4all_prof_add("update_fetch");		// 8
	uae4all_prof_add("linetoscr");			// 9
	uae4all_prof_add("do_long_fetch");		// 10
	uae4all_prof_add("pfield_doline");		// 11
	uae4all_prof_add("draw_sprites_ecs");		// 12
	uae4all_prof_add("flush_block");		// 13
	uae4all_prof_add("SET_INTERRUPT");		// 14
/*
	uae4all_prof_add("15");		// 15
	uae4all_prof_add("16");		// 16
	uae4all_prof_add("17");		// 17
	uae4all_prof_add("18");		// 18
	uae4all_prof_add("19");		// 19
	uae4all_prof_add("20");		// 20
	uae4all_prof_add("21");		// 21
	uae4all_prof_add("22");		// 22
*/
#endif
#ifdef DREAMCAST
	SDL_DC_EmulateKeyboard(SDL_FALSE);
#endif
	return 0;
    }
    return -1;
}

#ifdef DINGOO
static void show_mhz(void)
{
    extern int dingoo_get_clock(void);
    char n[40];
    sprintf((char *)&n[0],"Dingoo at %iMHz",dingoo_get_clock());
    gui_set_message((char *)&n[0], 50);
}

static void show_brightness(void)
{
    extern int dingoo_get_brightness(void);
    char n[40];
    sprintf((char *)&n[0],"Brightness %i%%",dingoo_get_brightness());
    gui_set_message((char *)&n[0], 40);
}

static void show_volumen(void)
{
    extern int dingoo_get_volumen(void);
    char n[40];
    sprintf((char *)&n[0],"Volumen %i%%",dingoo_get_volumen());
    gui_set_message((char *)&n[0], 40);
}

static void inc_dingoo_mhz(void)
{
	extern void dingoo_set_clock(unsigned int);
	extern unsigned int dingoo_get_clock(void);
	dingoo_set_clock(dingoo_get_clock()+5);
	show_mhz();
}

static void dec_dingoo_mhz(void)
{
	extern void dingoo_set_clock(unsigned int);
	extern unsigned int dingoo_get_clock(void);
	dingoo_set_clock(dingoo_get_clock()-5);
	show_mhz();
}

static void inc_dingoo_brightness(void)
{
	extern void dingoo_set_brightness(int);
	extern int dingoo_get_brightness(void);
	dingoo_set_brightness(dingoo_get_brightness()+2);
	show_brightness();
}

static void dec_dingoo_brightness(void)
{
	extern void dingoo_set_brightness(int);
	extern int dingoo_get_brightness(void);
	dingoo_set_brightness(dingoo_get_brightness()-2);
	show_brightness();
}
static void inc_dingoo_volumen(void)
{
	extern void dingoo_set_volumen(int);
	extern int dingoo_get_volumen(void);
	dingoo_set_volumen(dingoo_get_volumen()+1);
	show_volumen();
}

static void dec_dingoo_volumen(void)
{
	extern void dingoo_set_volumen(int);
	extern int dingoo_get_volumen(void);
	dingoo_set_volumen(dingoo_get_volumen()-1);
	show_volumen();
}

#else
#define show_mhz()
#define inc_dingoo_mhz()
#define dec_dingoo_mhz()
#define inc_dingoo_brightness()
#define dec_dingoo_brightness()
#define inc_dingoo_volumen()
#define dec_dingoo_volumen()
#endif

int gui_update (void)
{
    extern char *savestate_filename;
    extern int saveMenu_n_savestate;
// SE EJECUTA DESPUES DE INICIAR EL CORE 68k
    strcpy(changed_df[0],uae4all_image_file0);
    strcpy(changed_df[1],uae4all_image_file1);
    strcpy(changed_df[2],uae4all_image_file2);
    strcpy(changed_df[3],uae4all_image_file3);
    strcpy(savestate_filename,uae4all_image_file0);
    switch(saveMenu_n_savestate)
    {
	    case 1:
    		strcat(savestate_filename,"-1.asf");
	    case 2:
    		strcat(savestate_filename,"-2.asf");
	    case 3:
    		strcat(savestate_filename,"-3.asf");
	    default: 
    	   	strcat(savestate_filename,".asf");
    }
    real_changed_df[0]=1;
    real_changed_df[1]=1;
    real_changed_df[2]=1;
    real_changed_df[3]=1;
    show_mhz();
    return 0;
}


static void goMenu(void)
{
   int exitmode=0;
   int autosave=mainMenu_autosave;
   if (quit_program != 0)
	    return;
#ifdef PROFILER_UAE4ALL
   uae4all_prof_show();
#endif
#ifdef DEBUG_FRAMERATE
   uae4all_show_time();
#endif
   emulating=1;
//   vkbd_quit();
   init_text(0);
   pause_sound();
   menu_raise();
   exitmode=run_mainMenu();
   notice_screen_contents_lost();
   resume_sound();
   if ((!(strcmp(prefs_df[0],uae4all_image_file0))) || ((!(strcmp(prefs_df[1],uae4all_image_file1)))))
	   menu_unraise();
   quit_text();
//   vkbd_init();
#ifdef DREAMCAST
   SDL_DC_EmulateKeyboard(SDL_FALSE);
#endif
    getChanges();
    vkbd_init_button2();
    if (exitmode==1 || exitmode==2)
    {
    	    extern char *savestate_filename;
    	    extern int saveMenu_n_savestate;
	    for(int i=0;i<mainMenu_drives;i++)
	    {
		if (i==0 && strcmp(changed_df[0],uae4all_image_file0)) {
		    strcpy(changed_df[0],uae4all_image_file0);
		    real_changed_df[0]=1;
		}
		else if (i==1 && strcmp(changed_df[1],uae4all_image_file1)) {
		    strcpy(changed_df[1],uae4all_image_file1);
		    real_changed_df[1]=1;
		}
		else if (i==2 && strcmp(changed_df[2],uae4all_image_file2)) {
		    strcpy(changed_df[2],uae4all_image_file2);
		    real_changed_df[2]=1;
		}
		else if (i==3 && strcmp(changed_df[3],uae4all_image_file3)) {
		    strcpy(changed_df[3],uae4all_image_file3);
		    real_changed_df[3]=1;
		}
	    }
    	    strcpy(savestate_filename,uae4all_image_file0);
	    switch(saveMenu_n_savestate)
    	    {
	    	case 1:
    			strcat(savestate_filename,"-1.asf");
	    	case 2:
    			strcat(savestate_filename,"-2.asf");
	    	case 3:
    			strcat(savestate_filename,"-3.asf");
	    	default: 
    	   		strcat(savestate_filename,".asf");
    	    }
    }
    if (exitmode==3)
    {
    	    extern char *savestate_filename;
    	    extern int saveMenu_n_savestate;
	    for(int i=0;i<mainMenu_drives;i++)
	    {
		changed_df[i][0]=0;
		if (i==0) {
		    uae4all_image_file0[0]=0;
		    if (strcmp(changed_df[0],uae4all_image_file0))
		    { 
			strcpy(changed_df[0],uae4all_image_file0);
			real_changed_df[0]=1;
		    }
		}
		else if (i==1) {
		    uae4all_image_file1[0]=0;
		    if (strcmp(changed_df[1],uae4all_image_file1))
		    { 
			strcpy(changed_df[1],uae4all_image_file1);
			real_changed_df[1]=1;
		    }
		}
		else if (i==2) {
		    uae4all_image_file2[0]=0;
		    if (strcmp(changed_df[2],uae4all_image_file2))
		    { 
			strcpy(changed_df[2],uae4all_image_file2);
			real_changed_df[2]=1;
		    }
		}
		else if (i==3) {
		    uae4all_image_file3[0]=0;
		    if (strcmp(changed_df[3],uae4all_image_file3))
		    { 
			strcpy(changed_df[3],uae4all_image_file3);
			real_changed_df[3]=1;
		    }
		}
		disk_eject(i);
	    }
    	    strcpy(savestate_filename,uae4all_image_file0);
    	    switch(saveMenu_n_savestate)
    	    {
	   	case 1:
    			strcat(savestate_filename,"-1.asf");
	    	case 2:
    			strcat(savestate_filename,"-2.asf");
	    	case 3:
    			strcat(savestate_filename,"-3.asf");
	    	default: 
    	  	 	strcat(savestate_filename,".asf");
    	    }
    }
    if (exitmode==2)
    {
	    if (autosave!=mainMenu_autosave)
	    {
	    	prefs_df[0][0]=0;
	   	prefs_df[1][0]=0;
	    	prefs_df[2][0]=0;
	   	prefs_df[3][0]=0;
	    }
	    black_screen_now();
	    show_mhz();
	    uae_reset ();
    }
    check_all_prefs();
    gui_purge_events();
    black_screen_now();
    notice_screen_contents_lost();
#ifdef DEBUG_FRAMERATE
    uae4all_update_time();
#endif
#ifdef PROFILER_UAE4ALL
    uae4all_prof_init();
#endif
}

static int nowSuperThrottle=0, goingSuperThrottle=0, goingVkbd=0, goingEmouse;

static void goSuperThrottle(void)
{
	if (!nowSuperThrottle)
	{
		nowSuperThrottle=1;
		m68k_speed=1; //6;
		changed_produce_sound=0;
		changed_gfx_framerate=80;
		check_prefs_changed_cpu();
		check_prefs_changed_audio();
		check_prefs_changed_custom();
		gui_set_message("SuperThrottle On",5);
	}
}

static void leftSuperThrottle(void)
{
	if (nowSuperThrottle)
	{
		nowSuperThrottle=0;
		getChanges();
		check_prefs_changed_cpu();
		check_prefs_changed_audio();
		check_prefs_changed_custom();
		gui_set_message("SuperThrottle Off",50);
	}
}

static void inc_throttle(int sgn)
{
	char n[40];
	static Uint32 last=0;
	Uint32 now=SDL_GetTicks();
	if (now-last<555)
		return;
	last=now;
	if (sgn>0)
	{
		if (mainMenu_throttle<5)
			mainMenu_throttle++;
	}
	else if (sgn<0)
		if (mainMenu_throttle>0)
			mainMenu_throttle--;
	nowSuperThrottle=0;
	getChanges();
	check_prefs_changed_cpu();
	check_prefs_changed_audio();
	check_prefs_changed_custom();
	sprintf((char *)&n[0],"Throttle %i",mainMenu_throttle*20);
	gui_set_message((char *)&n[0],50);
}

void gui_handle_events (void)
{
#ifndef DREAMCAST
	Uint8 *keystate = SDL_GetKeyState(NULL);

#ifdef DINGOO
	if (keystate[SDLK_ESCAPE])
	{
		if (keystate[SDLK_LCTRL])
		{
			keystate[SDLK_LCTRL]=0;
			inc_dingoo_mhz();
		}
		else if (keystate[SDLK_LALT])
		{
			keystate[SDLK_LALT]=0;
			dec_dingoo_mhz();
		}
		else if (keystate[SDLK_SPACE])
		{
			keystate[SDLK_SPACE]=0;
			inc_throttle(1);
		}
		else if (keystate[SDLK_LSHIFT])
		{
			keystate[SDLK_LSHIFT]=0;
			inc_throttle(-1);
		}
		else if (keystate[SDLK_UP])
		{
			keystate[SDLK_UP]=0;
			inc_dingoo_brightness();
		}
		else if (keystate[SDLK_DOWN])
		{
			keystate[SDLK_DOWN]=0;
			dec_dingoo_brightness();
		}
		else if (keystate[SDLK_RIGHT])
		{
			keystate[SDLK_RIGHT]=0;
			inc_dingoo_volumen();
		}
		else if (keystate[SDLK_LEFT])
		{
			keystate[SDLK_LEFT]=0;
			dec_dingoo_volumen();
		}
	}
	else
#endif
#ifdef EMULATED_JOYSTICK
	if (emulated_mouse)
	{
		if (keystate[SDLK_LEFT])
		{
			lastmx -= emulated_mouse_speed;
	    		newmousecounters = 1;
		}
		else
		if (keystate[SDLK_RIGHT])
		{
			lastmx += emulated_mouse_speed;
	    		newmousecounters = 1;
		}
		if (keystate[SDLK_UP])
		{
			lastmy -= emulated_mouse_speed;
	    		newmousecounters = 1;
		}
		else
		if (keystate[SDLK_DOWN])
		{
			lastmy += emulated_mouse_speed;
	    		newmousecounters = 1;
		}

		emulated_mouse_button1=keystate[SDLK_LSHIFT];
		emulated_mouse_button2=keystate[SDLK_SPACE];
	}
	else if (emulated_joystick)
	{
		emulated_left=keystate[SDLK_LEFT];
		emulated_right=keystate[SDLK_RIGHT];
		emulated_top=keystate[SDLK_UP];
		emulated_bot=keystate[SDLK_DOWN];

		emulated_button1=keystate[SDLK_LSHIFT];
		emulated_button2=keystate[SDLK_SPACE];
	}
#endif
#ifndef DINGOO
	if ( keystate[SDLK_PAGEUP] )
		goSuperThrottle();
	else
		leftSuperThrottle();
#endif
#ifdef MAEMO_CHANGES
	if ( keystate[SDLK_MODE] && !keystate[SDLK_LSHIFT] && !keystate[SDLK_LCTRL] && keystate[SDLK_KP_ENTER] )
		SDL_WM_ToggleFullScreen(prSDLScreen);
	else
	if ( keystate[SDLK_MODE] && !keystate[SDLK_LSHIFT] && !keystate[SDLK_LCTRL] && keystate[SDLK_BACKSPACE] )
#else
#if !defined(DINGOO) && !defined(DREAMCAST)
	if ( keystate[SDLK_F12] )
		SDL_WM_ToggleFullScreen(prSDLScreen);
	else
#endif
	if (( keystate[SDLK_F11] )
#ifdef DINGOO
			||((keystate[SDLK_RETURN])&&(keystate[SDLK_ESCAPE]))
#endif
	   )
#endif // MAEMO_CHANGES
#else
	if (SDL_JoystickGetButton(uae4all_joy0,3) || SDL_JoystickGetButton(uae4all_joy1,3) )
#endif
	{
#ifdef DINGOO
		keystate[SDLK_RETURN]=keystate[SDLK_ESCAPE]=keystate[SDLK_TAB]=keystate[SDLK_BACKSPACE]=0;
#endif
		goMenu();
	}
#ifdef DINGOO
	if (keystate[SDLK_RETURN])
	{
		if (goingSuperThrottle<3)
			goingSuperThrottle++;
		else
		{
			goingSuperThrottle=-1000;
			if (nowSuperThrottle)
				leftSuperThrottle();
			else
				goSuperThrottle();
		}
	}
	else
		goingSuperThrottle=0;
	if (keystate[SDLK_TAB])
	{
		if (keystate[SDLK_ESCAPE])
		{
			keystate[SDLK_ESCAPE]=keystate[SDLK_TAB]=0;
			savestate_state = STATE_DOSAVE;
		}
		else
		if (!vkbd_mode)
			goingEmouse=1;
	}
	else if (goingEmouse)
	{
		goingEmouse=0;
		if (emulated_mouse)
   			notice_screen_contents_lost();
		emulated_mouse=~emulated_mouse;
	}
	if (keystate[SDLK_BACKSPACE])
	{
		if (keystate[SDLK_ESCAPE])
		{
			extern char *savestate_filename;
			FILE *f=fopen(savestate_filename,"rb");
			keystate[SDLK_ESCAPE]=keystate[SDLK_BACKSPACE]=0;
			if (f)
			{
				fclose(f);
				savestate_state = STATE_DORESTORE;
			}
			else
    				gui_set_message("Failed: Savestate not found", 100);
		}
		else
		if (!emulated_mouse)
			goingVkbd=1;
		else
		{
			char str[40];
			static Uint32 last=0;
			Uint32 now=SDL_GetTicks();
			if (now-last>100)
			{
				if (emulated_mouse_speed==15)
					emulated_mouse_speed=1;
				else
					emulated_mouse_speed++;
				sprintf((char *)&str[0],"  Mouse %i speed",emulated_mouse_speed);
    				gui_set_message((char *)&str[0], 40);
				last=now;
			}
		}
	}
	else if (goingVkbd)
	{
		goingVkbd=0;
		if (vkbd_mode)
		{
			vkbd_mode=0;
   			notice_screen_contents_lost();
		}
		else
			vkbd_mode=1;
	}
#endif
#ifdef DREAMCAST
	if (SDL_JoystickGetAxis (uae4all_joy0, 2))
	{
		if (vkbd_mode)
		{
			vkbd_mode=0;
			goingVkbd=0;
   			notice_screen_contents_lost();
		}
		else
			goingSuperThrottle=1;
	}
	else
		if (!nowSuperThrottle)
			goingSuperThrottle=0;
		else
			goingVkbd=0;
	if (SDL_JoystickGetAxis (uae4all_joy0, 3))
	{
		if (goingSuperThrottle)
			goSuperThrottle();
		else
			if (goingVkbd>4)
				vkbd_mode=1;
			else
				goingVkbd++;
	}
	else
		if (nowSuperThrottle)
			leftSuperThrottle();
		else
			goingVkbd=0;
#endif
	if (vkbd_key)
	{
		if (vkbd_keysave==-1234567)
		{
			SDL_keysym ks;
			ks.sym=vkbd_key;
			vkbd_keysave=keycode2amiga(&ks);
			if (vkbd_keysave >= 0)
			{
				if (!uae4all_keystate[vkbd_keysave])
				{
					uae4all_keystate[vkbd_keysave]=1;
					record_key(vkbd_keysave<<1);
				}
			}
		}
	}
	else
		if (vkbd_keysave!=-1234567)
		{
			if (vkbd_keysave >= 0)
			{
				uae4all_keystate[vkbd_keysave]=0;
				record_key((vkbd_keysave << 1) | 1);
			}
			vkbd_keysave=-1234567;
		}
}

void gui_changesettings (void)
{
	dbg("GUI: gui_changesettings");
}

void gui_update_gfx (void)
{
// ANTES DE LA ENTRADA EN VIDEO
//	dbg("GUI: gui_update_gfx");
}

void gui_set_message(char *msg, int t)
{
	show_message=t;
	strncpy(show_message_str, msg, 36);
}

void gui_show_window_bar(int per, int max, int case_title)
{
	char *title;
	if (case_title)
		title="  Restore State";
	else
		title="  Save State";
	_text_draw_window_bar(prSDLScreen,80,64,172,48,per,max,title);
#if defined(DOUBLEBUFFER) || defined(DINGOO)
	SDL_Flip(prSDLScreen);
#endif
}

/*
int run_menuGame() {}
int run_menuControl() {}
*/


#ifdef PROFILER_UAE4ALL

static unsigned uae4all_prof_total_initial=0;
unsigned uae4all_prof_total=0;
static char *uae4all_prof_msg[UAE4ALL_PROFILER_MAX];

void uae4all_prof_init(void)
{
	unsigned i;
#ifndef DREAMCAST
	unsigned long long s=SDL_GetTicks();
#else
	unsigned long long s=timer_us_gettime64();
#endif
	for(i=0;i<UAE4ALL_PROFILER_MAX;i++)
	{
		uae4all_prof_initial[i]=s;
		uae4all_prof_sum[i]=0;
		uae4all_prof_executed[i]=0;
		if (!uae4all_prof_total)
			uae4all_prof_msg[i]=NULL;
	}
	uae4all_prof_total_initial=s;
}

void uae4all_prof_add(char *msg)
{
	if (uae4all_prof_total<UAE4ALL_PROFILER_MAX)
	{
		uae4all_prof_msg[uae4all_prof_total]=msg;	
		uae4all_prof_total++;
	}
}

void uae4all_prof_show(void)
{
	unsigned i;
	double toper=0;
#ifndef DREAMCAST
	unsigned long long to=SDL_GetTicks()-uae4all_prof_total_initial;
#else
	unsigned long long to=uae4all_prof_sum[0]+uae4all_prof_sum[1];
	for(i=0;i<uae4all_prof_total;i++)
		if (uae4all_prof_sum[i]>to)
			uae4all_prof_sum[i]=0;
#endif

	puts("\n\n\n\n");
	puts("--------------------------------------------");
	for(i=0;i<uae4all_prof_total;i++)
	{
		unsigned long long t0=uae4all_prof_sum[i];
		double percent=(double)t0;
		percent*=100.0;
		percent/=(double)to;
		toper+=percent;
#ifdef DREAMCAST
		t0/=1000;
#endif
		printf("%s: %.2f%% -> Ticks=%i -> %iK veces\n",uae4all_prof_msg[i],percent,((unsigned)t0),(unsigned)(uae4all_prof_executed[i]>>10));
	}
	printf("TOTAL: %.2f%% -> Ticks=%i\n",toper,to);
	puts("--------------------------------------------"); fflush(stdout);
}
#endif
